### PR TITLE
refactor(frontend): generalize certify function for address

### PR DESCRIPTION
**Pull request by AntonioVentilii-DFINITY**
# Motivation

We generalize the `certifyAddress` function to be re-used for BTC.

### CODE REVIEW

File: src/frontend/src/lib/services/address.services.ts [hunk 1]
```diff
@@ -17,6 +18,8 @@ import type { OptionIdentity } from '$lib/types/identity';
 import type { TokenId } from '$lib/types/token';
 import { replacePlaceholders } from '$lib/utils/i18n.utils';
 import type { BitcoinNetwork } from '@dfinity/ckbtc';
+import type { address } from '@dfinity/ckbtc/dist/candid/bitcoin';
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/address.services.ts:21]: This type is lowercase?

File: src/frontend/src/lib/services/address.services.ts [hunk 2]
```diff
@@ -143,11 +146,17 @@ export const loadIdbAddress = async (): Promise<{ success: boolean }> => {
 	return { success: true };
 };
 
-export const certifyAddress = async (
-	address: string
-): Promise<{ success: boolean; err?: string }> => {
-	const tokenId = ETHEREUM_TOKEN_ID;
-
+const certifyAddress = async <T extends address>({
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/address.services.ts:149]: Not sure why `extends address` if this function is generic? Eth address does not extends adress or this is just for BTC?

File: src/frontend/src/lib/services/address.services.ts [hunk 3]
```diff
@@ -157,23 +166,47 @@ export const certifyAddress = async (
...
 		}
 
-		const certifiedAddress = await getEthAddress(identity);
+		const certifiedAddress = await getAddress(identity);
 
 		if (address.toLowerCase() !== certifiedAddress.toLowerCase()) {
 			return {
 				success: false,
-				err: 'The address used to load the data did not match your actual wallet address, which is why your session was ended. Please sign in again to reload your own data.'
+				err: `The address used to load the data did not match your actual ${tokenId.description} wallet address, which is why your session was ended. Please sign in again to reload your own data.`
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/address.services.ts:174]: I guess it's already the plan but, just in case, could you extract this i18n in another PR?

File: src/frontend/src/lib/services/address.services.ts [hunk 4]
```diff
@@ -157,23 +166,47 @@ export const certifyAddress = async (
...
 
 		addressStore.set({ tokenId, data: { data: address, certified: true } });
 
-		await updateIdbEthAddressLastUsage(identity.getPrincipal());
+		await updateIdbAddressLastUsage(identity.getPrincipal());
 	} catch (err: unknown) {
 		addressStore.reset(tokenId);
 
-		return { success: false, err: 'Error while loading the ETH address.' };
+		return { success: false, err: `Error while loading the ${tokenId.description} address.` };
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/address.services.ts:184]: same i18n key for error msg

File: src/frontend/src/lib/services/address.services.ts (hunk 2 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/services/address.services.ts]: no, you are right, it can be without extending, since it is generic

File: src/frontend/src/lib/services/address.services.ts (hunk 3 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/services/address.services.ts:173]: indeed I was going to do it in another PR, since there are actually more eheh

File: src/frontend/src/lib/services/address.services.ts (hunk 4 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/services/address.services.ts:183]: ditto

File: src/frontend/src/lib/services/address.services.ts [hunk 5]
```diff
@@ -179,23 +187,43 @@ export const certifyAddress = async (address: string): Promise<ResultSuccess<str
...
 	}
 
 	return { success: true };
 };
+
+export const certifyBtcAddressMainnet = async (address: string): Promise<ResultSuccess<string>> =>
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/address.services.ts:211]: It's not type `Address`? or that's in another PR?

File: src/frontend/src/lib/services/address.services.ts [hunk 6]
```diff
@@ -179,23 +187,43 @@ export const certifyAddress = async (address: string): Promise<ResultSuccess<str
...
+	certifyAddress({
+		tokenId: BTC_MAINNET_TOKEN_ID,
+		address,
+		getAddress: (identity) =>
```
Comment (peterpeterparker) [inline src/frontend/src/lib/services/address.services.ts:215]: `identity` has no type here? Typescript inherits it?

**Review by peterpeterparker — APPROVED** [review]
LGTM, thanks

Minor observations

File: src/frontend/src/lib/services/address.services.ts (hunk 5 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/services/address.services.ts:211]: right! good catch!

File: src/frontend/src/lib/services/address.services.ts (hunk 6 above)
Comment (AntonioVentilii-DFINITY) [reply src/frontend/src/lib/services/address.services.ts:215]: indeed it inherits it, that is why it didn't raise an error. But I will put it explicits just to be sure
